### PR TITLE
cleanup: clean up of GCOVR macros

### PR DESCRIPTION
**Pull request by alyssawilk**
Removing NOT_REACHED_GCOVR_EXCL_LINE which I think are extraneous or replacing with ENVOY_BUG where possible. 

Brings envoy GCOVR based PANICs from 358 -> 292

Risk Level: low
Testing: existing tests
Docs Changes: n/a
Release Notes: n/a

Part of https://github.com/envoyproxy/envoy/issues/19172

### CODE REVIEW

File: source/common/common/assert.h [hunk 1]
```diff
@@ -247,6 +247,8 @@ void resetEnvoyBugCountersForTest();
  */
 #define ENVOY_BUG(...) PASS_ON(PASS_ON(_ENVOY_BUG_VERBOSE)(__VA_ARGS__))
 
+#define IS_ENVOY_BUG(...) ENVOY_BUG(false, __VA_ARGS__);
```
Comment (asraa) [inline source/common/common/assert.h:250]: Could you please add a small comment on its usage ("Always triggering ENVOY_BUG intended for paths that are not expected to be reached.")?

**Review by asraa — CHANGES_REQUESTED** [review]
Thanks!

/wait

**Comment by wrowe** [comment]
This is looking solid but conflicts need to be resolved, please merge main @alyssawilk

File: source/common/common/matchers.cc [hunk 2]
```diff
@@ -32,7 +32,7 @@ ValueMatcherConstSharedPtr ValueMatcher::create(const envoy::type::matcher::v3::
   case envoy::type::matcher::v3::ValueMatcher::MatchPatternCase::kListMatch:
     return std::make_shared<const ListMatcher>(v.list_match());
   default:
-    NOT_REACHED_GCOVR_EXCL_LINE;
+    throw EnvoyException("Uncaught default");
```
Comment (jmarantz) [inline source/common/common/matchers.cc:35]: IIUC, `NOT_REACHED_GCOVR_EXCL_LINE` was originally a signal to the coverage system not to count that line against us in the line-coverage computation. It was also a signal to humans that was the intent.

At some point our coverage infrastructure changed, and `NOT_REACHED_GCOVR_EXCL_LINE` no longer is recognized as a "this doesn't count" signal to the coverage metrics, but it is still a useful signal (IMO) to humans indicating we don't expect the line to be hit, and someday maybe we'll be able to convince the coverage system, once again, not to count such lines.

I feel like IS_ENVOY_BUG and this `throw` are less clear signals for that. On the flip side, your change adds more details in many cases which might be useful to the human reader.

But I wonder if we could keep some kind of coverage-specific macro indicating the intent, with the detail you've added, even if isn't actually recognized (yet) by any of our coverage tools.

WDYT?

Comment (alyssawilk) [reply source/common/common/matchers.cc:35]: I think this was already discussed on slack and this is what we came up with :-)

I think also saying "it's a bug to hit this line of code" is a pretty clear indicator that code should not be hit.  If we ever have coverage tools which allow excluding and don't have that name we can always bulk rename then?

File: source/server/admin/admin.h [hunk 3]
```diff
@@ -232,7 +232,9 @@ class AdminImpl : public Admin,
     SystemTime lastUpdated() const override { return time_source_.systemTime(); }
     void onConfigUpdate() override {}
     void requestVirtualHostsUpdate(const std::string&, Event::Dispatcher&,
-                                   std::weak_ptr<Http::RouteConfigUpdatedCallback>) override {}
+                                   std::weak_ptr<Http::RouteConfigUpdatedCallback>) override {
+      IS_ENVOY_BUG("Unexpected function call");
```
Comment (jmarantz) [inline source/server/admin/admin.h:236]: how did you decide to add this new assertion?

File: source/common/tcp_proxy/tcp_proxy.cc [hunk 4]
```diff
@@ -444,18 +444,13 @@ void Filter::onGenericPoolFailure(ConnectionPool::PoolFailureReason reason,
   case ConnectionPool::PoolFailureReason::Overflow:
   case ConnectionPool::PoolFailureReason::LocalConnectionFailure:
     upstream_callbacks_->onEvent(Network::ConnectionEvent::LocalClose);
-    break;
-
+    return;
```
Comment (jmarantz) [inline source/common/tcp_proxy/tcp_proxy.cc:447]: why is return better than break here?

**Review by jmarantz — COMMENTED** [review]
OK -- that's fine. Is it possible to link to the slack thread in the description?

Just a couple small comments from skimming through the code changes.

**Comment by alyssawilk** [comment]
the PRs should be able to land in either order or in parallel but I'll focus on this one just to get one to land.

**Comment by alyssawilk** [comment]
slack thread was on #maintainers so I don't think linking would help

**Review by jmarantz — APPROVED** [review]
Thanks! If you want you can summarize the slack thread in the description so we can see the context which motivated this change.

**Comment by alyssawilk** [comment]
I think most of the discussion on the linked issue but I've added a tl;dr at the top to make it more clear what the plan is in why.  lmk if you think there's gaps?

**Comment by jmarantz** [comment]
oh I had never clicked thru that; my bad. Feel free to merge. thanks.

**Pull request merged**